ouroboros-consensus-protocol/src/Ouroboros/Consensus/Protocol/Praos.hs

```diff
@@ -0,0 +1,684 @@
+
+instance PraosCrypto c => NoThunks (ConsensusConfig (Praos c))
+
+type PraosValidateView c = Views.HeaderView c
+
+{-------------------------------------------------------------------------------
+  ConsensusProtocol
+-------------------------------------------------------------------------------}
+
+-- | Ledger view at a particular slot
+newtype instance Ticked (Views.LedgerView c) = TickedPraosLedgerView
+  { -- TODO: Perhaps it would be cleaner to define this as a separate type
+    getTickedPraosLedgerView :: Views.LedgerView c
+  }
+
+-- | Transitional Praos consensus state.
+--
+-- In addition to the 'ChainDepState' provided by the ledger, we track the slot
+-- number of the last applied header.
+data PraosState c = PraosState
+  { praosStateLastSlot            :: !(WithOrigin SlotNo),
+    -- | Operation Certificate counters
+    praosStateOCertCounters       :: !(Map (KeyHash 'BlockIssuer c) Word64),
+    -- | Evolving nonce
+    praosStateEvolvingNonce       :: !Nonce,
+    -- | Candidate nonce
+    praosStateCandidateNonce      :: !Nonce,
+    -- | Epoch nonce
+    praosStateEpochNonce          :: !Nonce,
+    -- | Nonce constructed from the hash of the previous block
+    praosStateLabNonce            :: !Nonce,
+    -- | Nonce corresponding to the LAB nonce of the last block of the previous
+    -- epoch
+    praosStateLastEpochBlockNonce :: !Nonce
+  }
+  deriving (Generic, Show, Eq)
+
+instance PraosCrypto c => NoThunks (PraosState c)
+
+-- | Version 0 supported rollback, removed in #2575.
+serialisationFormatVersion1 :: VersionNumber
+serialisationFormatVersion1 = 1
+
+instance PraosCrypto c => Serialise (PraosState c) where
+  encode
+    PraosState
+      { praosStateLastSlot,
+        praosStateOCertCounters,
+        praosStateEvolvingNonce,
+        praosStateCandidateNonce,
+        praosStateEpochNonce,
+        praosStateLabNonce,
+        praosStateLastEpochBlockNonce
+      } =
+      encodeVersion serialisationFormatVersion1 $
+        mconcat
+          [ CBOR.encodeListLen 2,
+            toCBOR praosStateLastSlot,
+            toCBOR praosStateOCertCounters,
+            toCBOR praosStateEvolvingNonce,
+            toCBOR praosStateCandidateNonce,
+            toCBOR praosStateEpochNonce,
+            toCBOR praosStateLabNonce,
+            toCBOR praosStateLastEpochBlockNonce
+          ]
+
+  decode =
+    decodeVersion
+      [(serialisationFormatVersion1, Decode decodePraosState1)]
+    where
+      decodePraosState1 = do
+        enforceSize "PraosState" 7
+        PraosState
+          <$> fromCBOR
+          <*> fromCBOR
+          <*> fromCBOR
+          <*> fromCBOR
+          <*> fromCBOR
+          <*> fromCBOR
+          <*> fromCBOR
+
+-- | Ticked 'PraosState'
+data instance Ticked (PraosState c) = TickedPraosState
+  { tickedPraosStateChainDepState :: PraosState c,
+    tickedPraosStateLedgerView :: Ticked (Views.LedgerView c)
+  }
+
+-- | Errors which we might encounter
+data PraosValidationErr c
```

Comment:
Ok, so this is supposed to correspond to `PrtclPredicateFailure` from the ledger, but expanded into a single flat set of failures, and omitting the failures related to the BFT schedule that can no longer occur.

The ledger types are nested. `PrtclPredicateFailure` contains `OverlayPredicateFailure` and `UpdnPredicateFailure` however the latter is not relevant for protocol failures. The `OverlayPredicateFailure` also contains `OcertPredicateFailure`.

This does indeed contain all of the cases from `OcertPredicateFailure`, with one extra failure `CounterOverIncrementedOCERT` which is indeed appropriate for this protocol.

A difference vs ``OverlayPredicateFailure` is that it contains both `VRFKeyBadNonce` and `VRFKeyBadLeaderValue`, whereas this type only contains `VRFKeyBadNonce`. This makes some sense since the new Praos protocol uses a single VRF rather than a separate leader and nonce VRF. However this also suggests that we should rename the constructor here to reflect the fact that it's a unified VRF, rather than having the name follow just one of the two (the nonce).

I don't have a concrete suggestion for the name, but we should find something suitable and try and be consistent. Perhaps since there is now just a single VRF we can just call it _the_ VRF. So that'd suggest `VRFKeyBad`, and leave it at that.

Reply:
Gone for `VRFKeyBadProof`